### PR TITLE
Add support for creating a domain using a Virtualmin CLI-like way

### PR DESCRIPTION
**Pull request by iliajie**
Hello Jamie!

It was fairly difficult to implement, but I have created a clean way to set up a domain using a simple API, as discussed earlier.

Example of usage:

```Perl
foreign_require('virtual-server');
virtual_server::load_api('create-domain');
my $dom = virtual_server::create_domain_cli({
    domain => 'domain1.example.tld',
    pass => '123',
    unix => 1,
    dir => 1,
    dns => 1,
    web => 1,
    ssl => 1
});
```

This API is fully compatible with the Virtualmin CLI `create-domain` command.

### CODE REVIEW

**Comment by iliajie** [comment]
And this is how we will use the new API https://github.com/virtualmin/virtualmin-pro/commit/ef769bfe06a6e9a5dffaa9144d4588789b21d20d.

**Comment by iliajie** [comment]
Later on, as a logical continuation of this, we’ll be able to do more similar factoring, eventually droping all the logic from related sub-programs like `create-domain`. We can then just pass all the parsed parameters to the new sub, like this:

```perl
#!/usr/local/bin/perl

=head1 create-domain.pl

Create a virtual server
...

=cut

my $flags = &parse_cli_args();
eval { &create_domain_cli($flags); };
if ($@) {
  usage();
}
```

However, that will be a later and unrelated step, as I’m still working on PR https://github.com/virtualmin/virtualmin-gpl/pull/913 to move all flags from the existing `usage()` function into a unique list (hash) of flags for each sub-program. This will allow us to dynamically generate usage information and test flags.

**Comment by jcameron** [comment]
I'm sorry but this seems like the wrong approach. It's like you took the existing `create-domain.pl` and then just turned all args into params to this new function. But that adds an unnecessary extra intermediate format that's not at all needed.

What I think would be much cleaner is an API like :

```
$overrides = {
  'owner' => 'Bob Smith',
  'parent' => $parent->{'id'},
};
create_simple_virtual_server($domain, $overrides);
```

The `create_simple_virtual_server` function would do all validation, selection of features, etc. But callers can override some domain attributes with the `$overrides` hash ref, for example to set features or choose the type of domain.

**Comment by iliajie** [comment]
> I'm sorry but this seems like the wrong approach.

🙂 How can this be a wrong approach?

> It's like you took the existing `create-domain.pl` and then just turned all args into params to this new function.

Yes, exactly! The `create-domain.pl` implementation you made for domain setup is solid and already in place — no need to redo it. It’s been tested and works great. If any bugs pop up later, we’ll only need to fix this subroutine since it will cover both CLI and UI tasks. It also handles variable scoping and text abstraction, and runs in `strict` mode.

If a bug shows up in the future, this subroutine will be the only place we need to check.

> But that adds an unnecessary extra intermediate format that's not at all needed.

You probably missed my earlier comment, but this subroutine _can/will_ also be used by `create-domain.pl`.


> But callers can override some domain attributes with the $overrides hash ref, for example to set features or choose the type of domain.

Yes. I’m already doing that, but I’m not passing the domain name in a separate variable — I’ve just fixed it!


> What I think would be much cleaner is an API like :
> 
> ```
> $overrides = {
>   'owner' => 'Bob Smith',
>   'parent' => $parent->{'id'},
> };
> create_simple_virtual_server($domain, $overrides);
> ```

Yes, I have just fixed it to pass the domain name as a separate parameter, e.g.:

```
my $overrides = {
    pass => '123',
    unix => 1,
    dir => 1,
    dns => 1,
    web => 1,
    ssl => 1
};
my $dom = create_domain_cli('domain1.example.tld', $overrides);
```

It will either return an error or a reference to the newly created domain object.

The last improvement could be adjusting `create_domain_cli` so it works without any overrides at all. If no plan ID is given, it could automatically enable the default minimal necessary features like `unix`, `dir`, `web`, and `ssl`. I’ll leave that up to you if you’d like to enhance it further, making it so calling `create_domain_cli('domain1.example.tld', { pass => '1234' });` would be enough to create a new domain. Even as it stands, it’s much simpler and works perfectly for what we need.

File: virtual-server-lib-funcs.pl [hunk 1]
```diff
@@ -21100,6 +21100,22 @@ sub validation_select_features
 return @fopts;
 }
 
+# load_api(lib)
+# Load given API
+sub load_api
```
Comment (jcameron) [inline virtual-server-lib-funcs.pl:21105]: This is nice! But please factor it out into a separate PR

File: lang/en [hunk 2]
```diff
@@ -8448,4 +8448,61 @@ monitor_nodoms=No domains to check found!
 monitor_nofeats=No features to check found!
 monitor_done=Successfully checked $2 features in $1 domains
 
+cli_create_domain_missing_domain_name=Missing domain name
```
Comment (jcameron) [inline lang/en:8451]: The whole "cli_create_domain_" prefix is too long, and also misleading since this isn't necesarily only used by CLI commands. Maybe replace it with "newdomain_" instead?

Comment (iliajie) [reply lang/en:8451]: Yes, agreed! Will do!

File: virtual-server-lib-funcs.pl (hunk 1 above)
Comment (iliajie) [reply virtual-server-lib-funcs.pl:21105]: But why? This change introduced the feature and is directly connected to it...

File: lang/en (hunk 2 above)
Comment (iliajie) [reply lang/en]: Fixed now. I used even shorter `api_ndom_`. Here is the patch https://github.com/virtualmin/virtualmin-gpl/pull/917/commits/4cd1e6defb951c4e57331884da1e7b2f3449bfad.

Comment (jcameron) [reply lang/en]: You could even drop the `api_` from the string!

File: create-domain-api.pl [hunk 3]
```diff
@@ -0,0 +1,1026 @@
...
+
+if (defined($opts->{'email'})) {
+	$email = $opts->{'email'};
+	&extract_address_parts($email) ||
+		&error($text{'api_ndom_invalid_email_address'});
+	}
+
+my $user  = defined($opts->{'user'})  ? lc($opts->{'user'})  : undef;
+my $group = defined($opts->{'group'}) ? lc($opts->{'group'}) : undef;
+my $pass;
```
Comment (jcameron) [inline create-domain-api.pl:48]: Why not just assign to `$pass` without the `if` statement?

File: create-domain-api.pl [hunk 4]
```diff
@@ -0,0 +1,1026 @@
...
+	$pass = $opts->{'pass'};
+	}
+
+if (defined($opts->{'passfile'})) {
+	$pass = &read_file_contents($opts->{'passfile'});
+	$pass =~ s/\r|\n//g;
+	}
+
+my ($mysqlpass, $postgrespass, $hashpass);
+$mysqlpass    = $opts->{'mysql-pass'}    if (defined($opts->{'mysql-pass'}));
```
Comment (jcameron) [inline create-domain-api.pl:59]: Same here, no need for `if defined`

File: create-domain-api.pl [hunk 5]
```diff
@@ -0,0 +1,1026 @@
...
+	$uquota = 0 if ($uquota =~ /unlimited/i);
+	$anylimits = 1;
+	}
+
+# Set features
+my %feature = ();
+foreach my $f (@features) {
+	if ($opts->{$f}) {
+		$config{$f} ||
+		    &error(&text('api_ndom_feature_not_enabled', $f));
```
Comment (jcameron) [inline create-domain-api.pl:81]: Wouldn't it be better if this API returned an error message instead of calling `error` ?

File: virtual-server-lib-funcs.pl (hunk 1 above)
Comment (jcameron) [reply virtual-server-lib-funcs.pl:21105]: Where is this `load_api` function called?

**Comment by jcameron** [comment]
In general, you need to remove most of those un-necessary uses of `if defined($opts->{'x'})` since it's simpler to just assign to the other variable regardless.

File: create-domain-api.pl (hunk 4 above)
Comment (iliajie) [reply create-domain-api.pl]: Yes, agreed! I have just fixed that!

File: create-domain-api.pl (hunk 5 above)
Comment (iliajie) [reply create-domain-api.pl:77]: Not really, because this subroutine (i.e. `create_domain_cli`) is meant to be called in both the UI and CLI. In the CLI, we want to locally override the standard error function, which is intended for UI, by using:
```perl
{
    local *error = sub {
        my $message = shift;
        # Convert HTML to plain text
        $message = html_tags_to_text($message);
        print "Error: $message\n";
        };
    create_domain_cli('example.tld', { 'pass' => 1234, unix => 1, dir => 1, web => 1 });
}
```

File: virtual-server-lib-funcs.pl (hunk 1 above)
Comment (iliajie) [reply virtual-server-lib-funcs.pl:21105]: > Where is this `load_api` function called?

It’s called when we need to use this API. In particular here https://github.com/virtualmin/virtualmin-pro/blob/master/scripts/workbench-lib.pl#L1040/

Comment (jcameron) [reply virtual-server-lib-funcs.pl:21105]: Oh .. but that's inconsistent with the rest of how Virtualmin's separate `*-lib.pl` files work. Why not just load this all the time, like we do here : https://github.com/virtualmin/virtualmin-gpl/blob/master/virtual-server-lib-funcs.pl#L19

File: create-domain-api.pl (hunk 5 above)
Comment (jcameron) [reply create-domain-api.pl:77]: No no, that's not the right approach. API functions like this should always return either an error message, or maybe a hash ref with more details on the result.

Comment (iliajie) [reply create-domain-api.pl]: > No no, that's not the right approach. API functions like this should always return either an error message, or maybe a hash ref with more details on the result.

Oh, you’re absolutely right! My bad! It’s definitely better to return an error message rather than die with an HTML message in APIs. Though, on success, we return a hash for the newly created domain.

Fixed here https://github.com/virtualmin/virtualmin-gpl/pull/917/commits/ba1d4f9c0deff6c550a71be5adc10bb36bae0fa7.

File: virtual-server-lib-funcs.pl (hunk 1 above)
Comment (iliajie) [reply virtual-server-lib-funcs.pl:21105]: You’re right, it doesn’t match how the other lib files are loaded in Virtualmin. And, I knew that (how other libs are loaded in Virtualmin), but I made the choice on purpose. I think this API should only load when it’s needed since it’s 1000 lines long and there’s no point having it around all the time. If we did the same with other functions (like `create-user-api.pl` and others), loading them all unconditionally would just add a huge unnecessary overhead.

Comment (jcameron) [reply virtual-server-lib-funcs.pl:21105]: Right, but the advantage of loading them in https://github.com/virtualmin/virtualmin-gpl/blob/master/virtual-server-lib-funcs.pl#L19 is that they are cached in memory after Webmin is started..

Comment (iliajie) [reply virtual-server-lib-funcs.pl:21105]: Yes, caching is an advantage for always-needed APIs. For an API that’s only needed for specific tasks, I think lazy loading is a better approach.

File: lang/en (hunk 2 above)
Comment (iliajie) [reply lang/en]: > You could even drop the `api_` from the string!

Nah, I prefer to keep the `api_` prefix. I think it’s important, and it’s short too...

Comment (jcameron) [reply lang/en]: ok fine!

File: virtual-server-lib-funcs.pl (hunk 1 above)
Comment (jcameron) [reply virtual-server-lib-funcs.pl:21105]: Yeah but it's inconsistent with the rest of the code. Consistency makes things easier to understand, even if it's a tiny bit less efficient.

Comment (iliajie) [reply virtual-server-lib-funcs.pl:21105]: I completely agree about consistency, but I feel like 1000 lines of code is more than a “tiny bit.” If this is being referenced across 10 or 20 similar subroutines, would we still be seeing it as just a "small inefficiency"?

Comment (jcameron) [reply virtual-server-lib-funcs.pl:21105]: What efficiency are you concerned about though? The RAM consumed by the code when it's loaded? The cost of parsing it is a one-off at startup time.

Comment (iliajie) [reply virtual-server-lib-funcs.pl:21105]: Well, sure, we can do as you suggest, but for no better reason, we don’t load all the Perl modules in vendor_perl upfront — we load them on demand. So why not do it here, or conversely, why not load all the modules at once?

Is caching the only reason?

Comment (iliajie) [reply virtual-server-lib-funcs.pl:21105]: > Oh .. but that's inconsistent with the rest of how Virtualmin's separate `*-lib.pl` files work.

Fixed here https://github.com/virtualmin/virtualmin-gpl/pull/917/commits/8f101760dc6552bd6c0042e925e4f09ae3d096e5.

Comment (jcameron) [reply virtual-server-lib-funcs.pl:21105]: We do actually load several of the important Perl modiules up-front like POSIX and Socket, via miniserv.pl. In fact I'd be open to pre-loading more, as one of my big concerns in Webmin is page load latency due to the time spent parsing Perl..


Thanks for the fix!

Comment (iliajie) [reply virtual-server-lib-funcs.pl:21105]: > In fact I'd be open to pre-loading more, as one of my big concerns in Webmin is page load latency due to the time spent parsing Perl

We could, but the load depends directly on the available computing power, network speed, and latency.

I remember trying to preload theme files a couple of years ago, but something didn’t work as expected.

> Thanks for the fix!

You’re welcome! Is there anything else we need to fix here before merging this PR?

Comment (jcameron) [reply virtual-server-lib-funcs.pl:21105]: I still had a few open comments on this PR, like about the use of `if (defined($x)) {`

File: api-create-domain-lib.pl [hunk 6]
```diff
@@ -0,0 +1,1006 @@
...
+	&check_ipaddress($dns_ip) ||
+		return $text{'api_ndom_invalid_dns_ip'};
+	}
+
+if ($opts->{'no-dns-ip'}) {
+	$dns_ip = "";
+	}
+
+my $mailboxlimit;
+if (defined($opts->{'mailboxlimit'}) || defined($opts->{'max-mailboxes'})) {
```
Comment (jcameron) [inline api-create-domain-lib.pl:178]: Why two different opts that do the same thing?

File: api-create-domain-lib.pl [hunk 7]
```diff
@@ -0,0 +1,1006 @@
...
+my $letsencrypt;
+if ($opts->{'letsencrypt'} || $opts->{'acme'}) {
+	$letsencrypt = 1;
+	}
+
+if ($opts->{'letsencrypt-always'} || $opts->{'acme-always'}) {
+	$letsencrypt = 2;
+	}
+
+my $jail;
```
Comment (jcameron) [inline api-create-domain-lib.pl:304]: A ternerary expression would be more readable here

File: api-create-domain-lib.pl [hunk 8]
```diff
@@ -0,0 +1,1006 @@
...
+my $noslaves;
+$noslaves       = 1 if $opts->{'no-slaves'};
+my $nosecondaries;
+$nosecondaries  = 1 if $opts->{'no-secondaries'};
+
+my $precommand = $opts->{'pre-command'};
+my $postcommand = $opts->{'post-command'};
+
+my $letsencrypt;
+if ($opts->{'letsencrypt'} || $opts->{'acme'}) {
```
Comment (jcameron) [inline api-create-domain-lib.pl:296]: Again I don't see why two different opts are needed?

File: virtual-server-lib-funcs.pl (hunk 1 above)
Comment (jcameron) [reply virtual-server-lib-funcs.pl:21105]: I'm still seeing a lot of duplicated opts and use of `defined` that doesn't seem necessary. The flags to create-domain.pl shouldn't be just passed through to this new function.

Comment (iliajie) [reply virtual-server-lib-funcs.pl:21105]: > I'm still seeing a lot of duplicated opts and use of `defined` that doesn't seem necessary.

All of those left are necessary. If you think any of them aren’t, please let me know which in particular.

File: api-create-domain-lib.pl (hunk 6 above)
Comment (iliajie) [reply api-create-domain-lib.pl:178]: > Why two different opts that do the same thing?

This was in the original : https://github.com/virtualmin/virtualmin-gpl/blob/master/create-domain.pl#L262

File: api-create-domain-lib.pl (hunk 7 above)
Comment (iliajie) [reply api-create-domain-lib.pl]: Alright, fixed here https://github.com/virtualmin/virtualmin-gpl/pull/917/commits/43da2202bf5da3e4204201019a07542092a37f3c.

File: api-create-domain-lib.pl (hunk 8 above)
Comment (iliajie) [reply api-create-domain-lib.pl:296]: > Again I don't see why two different opts are needed?

Again, because it was in the original https://github.com/virtualmin/virtualmin-gpl/commit/a6dee79f8f61731760521a6710bdc6e77537a2a9#diff-2bac7886b5474f1bb1ec134dda768ff7aec92a80564c5f2a6eb12ba1f47f3a6aR358.

Comment (jcameron) [reply api-create-domain-lib.pl:296]: Right, but those are command line flags - because I wanted to rename the `--letsencrypt` flag to `--acme` , but wanted to keep compatibility with the old flag to not break other users of the API.  But since in this case you're creating a new API, there's no need to keep compatability.

Comment (iliajie) [reply api-create-domain-lib.pl]: Ah, yes, right! I fixed it here https://github.com/virtualmin/virtualmin-gpl/pull/917/commits/c45c7eb1dc5330505e9683009d899676746a84f8.

File: api-create-domain-lib.pl [hunk 9]
```diff
@@ -0,0 +1,999 @@
...
+
+my ($mysqlpass, $postgrespass, $hashpass);
+$mysqlpass    = $opts->{'mysql-pass'};
+$postgrespass = $opts->{'postgres-pass'};
+$hashpass     = 1 if ($opts->{'hashpass'});
+
+my ($quota, $uquota);
+if (defined($opts->{'quota'})) {
+	$quota = $opts->{'quota'};
+	$quota = 0 if ($quota =~ /unlimited/i);
```
Comment (jcameron) [inline api-create-domain-lib.pl:63]: Why not just expect the caller to pass in `0` meaning unlimited? It's an internal API, so no need to support human-friendly strings like `unlimited`

File: api-create-domain-lib.pl [hunk 10]
```diff
@@ -0,0 +1,999 @@
...
+	$anylimits = 1;
+	}
+
+if (defined($opts->{'uquota'})) {
+	$uquota = $opts->{'uquota'};
+	$uquota = 0 if ($uquota =~ /unlimited/i);
```
Comment (jcameron) [inline api-create-domain-lib.pl:69]: Same here

File: api-create-domain-lib.pl [hunk 11]
```diff
@@ -0,0 +1,999 @@
...
+	$ip = "allocate";    # will be done later
+	$virt = 1;
+	$name = 0;
+	}
+
+my $virtalready;
+$virtalready = 1 if ($opts->{'ip-already'});
+
+my $sharedip;
+if (defined($opts->{'shared-ip'})) {
```
Comment (jcameron) [inline api-create-domain-lib.pl:115]: No need for `defined`

File: api-create-domain-lib.pl [hunk 12]
```diff
@@ -0,0 +1,999 @@
...
+	# IPv6 on shared address
+	$ip6 = $opts->{'shared-ip6'};
+	$virt6 = 0;
+	$name6 = 1;
+	&indexof($ip6, &list_shared_ip6s()) >= 0 ||
+		return &text('api_ndom_ip6_not_in_shared_list', $ip6);
+	}
+
+my $dns_ip;
+if (defined($opts->{'dns-ip'})) {
```
Comment (jcameron) [inline api-create-domain-lib.pl:167]: No need for `defined`

File: api-create-domain-lib.pl (hunk 6 above)
Comment (jcameron) [reply api-create-domain-lib.pl:178]: Right, but again this is an internal API so there's no need to support legacy flag names

File: api-create-domain-lib.pl [hunk 13]
```diff
@@ -0,0 +1,999 @@
...
+	}
+
+my $mailboxlimit;
+if (defined($opts->{'mailboxlimit'}) || defined($opts->{'max-mailboxes'})) {
+	$mailboxlimit = $opts->{'mailboxlimit'} || $opts->{'max-mailboxes'};
+	$anylimits = 1;
+	}
+
+my $dbslimit;
+if (defined($opts->{'dbslimit'}) || defined($opts->{'max-dbs'})) {
```
Comment (jcameron) [inline api-create-domain-lib.pl:184]: Same here for the multiple option names

File: api-create-domain-lib.pl [hunk 14]
```diff
@@ -0,0 +1,999 @@
...
+	$dbslimit = $opts->{'dbslimit'} || $opts->{'max-dbs'};
+	$anylimits = 1;
+}
+
+my $domslimit;
+if (defined($opts->{'domslimit'}) || defined($opts->{'max-doms'})) {
```
Comment (jcameron) [inline api-create-domain-lib.pl:190]: Same here for the multiple option names

File: api-create-domain-lib.pl [hunk 15]
```diff
@@ -0,0 +1,999 @@
...
+	$domslimit = $opts->{'domslimit'} || $opts->{'max-doms'};
+	$anylimits = 1;
+	}
+
+my $aliaslimit;
+if (defined($opts->{'aliaslimit'}) || defined($opts->{'max-aliases'})) {
```
Comment (jcameron) [inline api-create-domain-lib.pl:196]: Same here for the multiple option names

File: api-create-domain-lib.pl [hunk 16]
```diff
@@ -0,0 +1,999 @@
...
+	$aliaslimit = $opts->{'aliaslimit'} || $opts->{'max-aliases'};
+	$anylimits = 1;
+	}
+
+my $aliasdomslimit;
+if (defined($opts->{'aliasdomslimit'}) || defined($opts->{'max-aliasdoms'})) {
```
Comment (jcameron) [inline api-create-domain-lib.pl:202]: Same here for the multiple option names

File: api-create-domain-lib.pl [hunk 17]
```diff
@@ -0,0 +1,999 @@
...
+	$aliasdomslimit = $opts->{'aliasdomslimit'} || $opts->{'max-aliasdoms'};
+	$anylimits = 1;
+	}
+
+my $realdomslimit;
+if (defined($opts->{'realdomslimit'}) || defined($opts->{'max-realdoms'})) {
```
Comment (jcameron) [inline api-create-domain-lib.pl:208]: Same here for the multiple option names

File: api-create-domain-lib.pl [hunk 18]
```diff
@@ -0,0 +1,999 @@
...
+	$anylimits = 1;
+	}
+
+my $tlimit;
+$tlimit = 1 if ($opts->{'limits-from-template'} || $opts->{'limits-from-plan'});
+
+my $prefix = $opts->{'prefix'} || $opts->{'suffix'};
+
+my $db;
+if (defined($opts->{'db'})) {
```
Comment (jcameron) [inline api-create-domain-lib.pl:249]: Is `defined` really needed?

File: api-create-domain-lib.pl [hunk 19]
```diff
@@ -0,0 +1,999 @@
...
+	$db = $opts->{'db'};
+	$db =~ /^[a-z0-9\-\_]+$/i || return $text{'invalid_database_name'};
+	}
+
+my $fwdto;
+if (defined($opts->{'fwdto'})) {
```
Comment (jcameron) [inline api-create-domain-lib.pl:255]: Is `defined` really needed?

File: api-create-domain-lib.pl (hunk 9 above)
Comment (iliajie) [reply api-create-domain-lib.pl:63]: We could, it’s not a problem. However, I’d like this function to be a drop-in replacement for the current code in `create-domain.pl`. If I change that now, I might forget later when building the `%opts` parameter for this new subroutine. Let's just support both, `unlimited` and `0`...

File: api-create-domain-lib.pl (hunk 13 above)
Comment (iliajie) [reply api-create-domain-lib.pl:184]: For this and all other "Same here for the multiple option names," as I mentioned [here](https://github.com/virtualmin/virtualmin-gpl/pull/917#discussion_r1805434186), I really want this new sub to be a _simple_ drop-in replacement for the existing code in `create-domain.pl` so we don’t end up supporting both.

File: api-create-domain-lib.pl (hunk 9 above)
Comment (jcameron) [reply api-create-domain-lib.pl:63]: I don't think that's good API design though. Internal APIs like this don't need to support human-friendly values, since we can just do the conversion from `unlimited` to `0` in `create-domain.pl`

File: api-create-domain-lib.pl (hunk 13 above)
Comment (jcameron) [reply api-create-domain-lib.pl:184]: It can't be a drop-in replacement because it's an internal API, but `create-domain` is an external API. It makes no sense to keep support for legacy flags in `create-domain` in a new internal API.

Comment (iliajie) [reply api-create-domain-lib.pl]: > It makes no sense to keep support for legacy flags in create-domain in a new internal API.

Agreed! Fixed on the latest commit!

**Comment by jcameron** [comment]
Ok we're looking a lot better now! Only major issue I see is the support for the "unlimited" quota value, which you could just move up to the caller.

**Comment by iliajie** [comment]
> Ok we're looking a lot better now! Only major issue I see is the support for the "unlimited" quota value, which you could just move up to the caller.

Alright! Fixed this too here https://github.com/virtualmin/virtualmin-gpl/pull/917/commits/cb29f69c1a6e40dd4dfe24d9b5d78d695f63b348.

Thanks!

File: virtual-server-lib-funcs.pl (hunk 1 above)
Comment (jcameron) [inline virtual-server-lib-funcs.pl:21105]: I guess we can get ride of this `load_api` function then?

Comment (iliajie) [reply virtual-server-lib-funcs.pl]: Ah, yes! Sorry about that! I thought I had, but I’ve just done it now.

**Comment by jcameron** [comment]
Thanks for the PR!

**Comment by iliajie** [comment]
> Thanks for the PR!

Thank _you_ for reviewing it and sharing your experience!
